src/string/is_valid_email_address.rs

```diff
@@ -0,0 +1,255 @@
+            return false;
+        }
+    }
+    true
+}
+
+fn contains_legal_domain_characters(domain: &str) -> bool {
+    for item in domain.chars() {
+        if !CHARACTERS.contains(item) && item != '-' && item != '.' && !"[]:".contains(item) {
+            return false;
+        }
+    }
+    true
+}
+
+fn has_valid_quotes(part: &str) -> bool {
+    for i in 1..part.len() {
+        if part.chars().nth(i - 1).unwrap() == '"' && part.chars().nth(i).unwrap() == '"' {
+            let proceeding_quote = part.chars().nth(i + 1);
+            if proceeding_quote.is_some() && proceeding_quote.unwrap() != '.' {
+                return false;
+            }
+        }
+    }
+    true
+}
+
+fn is_valid_ipv4(ip: &str) -> bool {
+    let parts = ip
+        .split('.')
+        .map(|x| x.to_string())
+        .collect::<Vec<String>>();
+    if parts.len() != 4 {
+        return false;
+    }
+
+    for value in parts {
+        let value = value.parse::<u32>();
+        if value.is_err() || value.unwrap() > 255 {
+            return false;
+        }
+    }
+
+    true
+}
+
+fn is_valid_ipv6(ip: &str) -> bool {
+    let parts = ip
+        .split(':')
+        .map(|x| x.to_string())
+        .collect::<Vec<String>>();
+
+    // we check for the length 9 here as the IPv6 counts as one of the parts
+    if parts.len() != 9 || parts.first().unwrap() != "IPv6" {
+        return false;
+    }
+
+    for value in &parts[1..parts.len()] {
+        if value.is_empty() || value.len() > 4 {
+            return false;
+        }
+
+        for chr in value.chars() {
+            if !HEX_CHARACTERS.contains(chr.to_ascii_lowercase()) {
+                return false;
+            }
+        }
+    }
+
+    true
+}
+
+fn has_bad_brackets(domain: &str) -> bool {
+    let (starts_with, ends_with) = (domain.starts_with('['), domain.ends_with(']'));
+    starts_with != ends_with
+}
+
+pub fn is_valid_ip(domain: &str) -> bool {
+    is_valid_ipv4(domain) || is_valid_ipv6(domain)
+}
+
+fn is_local_part_valid(local_part: &str) -> bool {
+    if local_part.len() > MAX_LOCAL_PART_LENGTH as usize {
+        return false;
+    }
+
+    if !has_valid_quotes(local_part) {
+        return false;
+    }
+
+    if !contains_legal_local_characters(local_part) {
+        return false;
+    }
+
+    if has_bad_dots(local_part) {
+        return false;
+    }
+
+    true
+}
+
+fn is_domain_valid(domain: &str) -> bool {
+    if domain.len() > MAX_DOMAIN_LENGTH as usize {
+        return false;
+    }
+
+    if !contains_legal_domain_characters(domain) {
+        return false;
+    }
+
+    if has_bad_dots(domain) {
+        return false;
+    }
+
+    if has_bad_brackets(domain) {
+        return false;
+    }
+
+    if domain.starts_with('[')
+        && domain.ends_with(']')
+        && !is_valid_ip(&domain[1..domain.len() - 1])
+    {
+        return false;
+    }
+
+    if domain.starts_with('-') || domain.ends_with('-') {
+        return false;
+    }
+
+    true
+}
```

Comment:
This could be simplified to
```rust
fn is_domain_valid(domain: &str) -> bool {
    if domain.starts_with('[') && domain.ends_with(']')
    {
        return is_valid_ip(&domain[1..domain.len() - 1])
    }

    if domain.len() > MAX_DOMAIN_LENGTH as usize {
        return false;
    }

    if !contains_legal_domain_characters(domain) {
        return false;
    }

    if has_bad_dots(domain) {
        return false;
    }

    if domain.starts_with('[') || domain.ends_with(']') {
        return false
    }

    if domain.starts_with('-') || domain.ends_with('-') {
        return false;
    }

    true
}
```
But it can be simplified even further: for example when you know that you are not dealing with the ip address, you know that a valid domain will not contain any bracket.

Reply:
I simplified the code as previously mentioned